src/core/labeling/rules/qgslabelingenginerule.h

```diff
@@ -0,0 +1,299 @@
+     * \see extent()
+     */
+    QgsGeometry mapBoundaryGeometry() const;
+
+    /**
+     * Sets the map label boundary \a geometry, which defines the limits within which labels may be placed
+     * in the map.
+     *
+     * The map boundary geometry specifies the actual geometry of the map
+     * boundary, which will be used to detect whether a label is visible (or partially visible) in
+     * the rendered map. This may differ from extent() in the case of rotated or non-rectangular
+     * maps.
+     *
+     * \see setExtent()
+     * \see mapBoundaryGeometry()
+     */
+    void setMapBoundaryGeometry( const QgsGeometry &geometry );
+
+  private:
+
+#ifdef SIP_RUN
+    QgsLabelingEngineContext( const QgsLabelingEngineContext &other );
+#endif
+
+    QgsRenderContext &mRenderContext;
+    QgsRectangle mExtent;
+    QgsGeometry mMapBoundaryGeometry;
+};
+
+/**
+ * Abstract base class for labeling engine rules.
+ *
+ * Labeling engine rules implement custom logic to modify the labeling solution for a map render,
+ * e.g. by preventing labels being placed which violate custom constraints.
+ *
+ * \note QgsAbstractLabelingEngineRule cannot be subclassed in Python. Use one of the existing
+ * implementations of this class instead.
+ *
+ * \ingroup core
+ * \since QGIS 3.40
+ */
+class CORE_EXPORT QgsAbstractLabelingEngineRule
+{
+
+#ifdef SIP_RUN
+    SIP_CONVERT_TO_SUBCLASS_CODE
+    if ( sipCpp->id() == "minimumDistanceLabelToFeature" )
+    {
+      sipType = sipType_QgsLabelingEngineRuleMinimumDistanceLabelToFeature;
+    }
+    else if ( sipCpp->id() == "minimumDistanceLabelToLabel" )
+    {
+      sipType = sipType_QgsLabelingEngineRuleMinimumDistanceLabelToLabel;
+    }
+    else if ( sipCpp->id() == "maximumDistanceLabelToFeature" )
+    {
+      sipType = sipType_QgsLabelingEngineRuleMaximumDistanceLabelToFeature;
+    }
+    else if ( sipCpp->id() == "avoidLabelOverlapWithFeature" )
+    {
+      sipType = sipType_QgsLabelingEngineRuleAvoidLabelOverlapWithFeature;
+    }
+    else
+    {
+      sipType = 0;
+    }
+    SIP_END
+#endif
+
+  public:
+
+    virtual ~QgsAbstractLabelingEngineRule();
+
+    /**
+     * Creates a clone of this rule.
+     *
+     * The caller takes ownership of the returned object.
+     */
+    virtual QgsAbstractLabelingEngineRule *clone() const = 0 SIP_FACTORY;
+
+    /**
+     * Returns a string uniquely identifying the rule subclass.
+     */
+    virtual QString id() const = 0;
+
+    /**
+     * Returns a user-friendly, translated string representing the rule type.
+     */
+    virtual QString displayType() const = 0;
+
+    /**
+     * Returns a user-friendly description of the rule.
+     *
+     * This should include the rule name() if set, and other useful details for users
+     * to quickly identify the rule's purpose when shown in a tooltip.
+     *
+     * The returned string may contain HTML formatting.
+     */
+    virtual QString description() const;
```

Comment:
An icon representation could be nice too (not blocking this PR for that, just food for thoughts).